src/runner.rs

```diff
@@ -217,6 +216,72 @@ pub fn run(av: &ArgValues) -> error::Result<()> {
             }
         });
 
+        // notify handling
+        shader_modified = shader_modified | {
+            let mut have_events = false;
+            let basename = shader_basename.unwrap();
+
+            loop {
+                match rx.try_recv() {
+                    Err(TryRecvError::Empty) => break,
+
+                    // we handle both create and write here because some text editors write
+                    // the modified file to a tmpfile then move it 
+                    Ok(DebouncedEvent::Create(ref path)) | Ok(DebouncedEvent::Write(ref path))
+                        | Ok(DebouncedEvent::Rename(_, ref path))
+                        if path.ends_with(basename) => have_events = true,
+
+                    Ok(_ev) => {
+                        // println!(" >> unhandled notify event: {:?}", _ev);
+                    },
+
+                    Err(TryRecvError::Disconnected) => {
+                        println!(" !! watch disconnected");
+                        break;
+                    }
+                }
+            }
+
+            have_events
+        };
+
+        // this is a while loop so we can "break;" out of it prematurely
+        // in the event that we can't recompile the shader, this means that the
+        // old shader just continues running, and then we dump the error to stdout.
+        while shader_modified {
```

Comment:
I'm no rust expert, but couldn't this be made cleaner by using a function that returns early in case the shader cannot be recompiled ?

Reply:
I agree, this could probably be made into a function call

Reply:
yeah absolutely. the issue is that the shader recompilation modifies a few locals (the pipeline and uniforms) and I honestly was just too lazy to factor those out.

previously, some of the hairiness was avoided due to the shader recompilation happening inside a closure passed to `poll_events()`, and instead of `break` it was just using `return`.

@fmenozzi the right move here is probably pulling locals out of `run()` and having some sort of `State` struct that manages the pipeline and uniforms and probably pulling out the initial pipeline creation into the same function (I would think). I could hopefully take a look at that sometime this week if you'd like.

Reply:
That'd be great! I've been meaning to see if I could clean up the runner code anyways, and this would be a good first step.